src/kaocha/testable.clj

```diff
@@ -212,22 +223,68 @@
         (run % test-plan)
         (plugin/run-hook :kaocha.hooks/post-test % test-plan)))))
 
-(defn run-testables
+(defn try-run-testable [test test-plan n]
+  (let [result (try (run-testable test test-plan) (catch Exception _e false))]
+    (if (or result (> n 1))
+      ;; success or last try, return
+      result
+      ;; otherwise retry
+      (try-run-testable test test-plan (dec n)))))
+
+(defn run-testables-serial
   "Run a collection of testables, returning a result collection."
   [testables test-plan]
   (let [load-error? (some ::load-error testables)]
-    (loop [result []
+    (loop [result  []
            [test & testables] testables]
       (if test
         (let [test (cond-> test
                      (and load-error? (not (::load-error test)))
                      (assoc ::skip true))
               r (run-testable test test-plan)]
           (if (or (and *fail-fast?* (result/failed? r)) (::skip-remaining? r))
-            (reduce into result [[r] testables])
+            (into (conj result r) testables)
             (recur (conj result r) testables)))
         result))))
 
+(defn current-thread-info []
+  (let [thread (Thread/currentThread)]
+    {:name (.getName thread)
+     :id (.getId thread)
+     :group-name (.getName (.getThreadGroup thread))}))
+
+(defn run-testables-parallel
+  "Run a collection of testables in parallel, returning a result collection."
+  [testables test-plan]
+  (let [load-error? (some ::load-error testables)
+        futures (doall
+                 (map (fn [t]
+                        (future
+                          (run-testable (assoc t ::thread-info (current-thread-info)) test-plan)))
+                      testables))]
+    (doall (map deref futures))))
```

Comment:
Here we are creating an unbounded number of threads. If the test plan contains too many tests (hundreds or thousands) this will probably be a problem.

OS threads are expensive to create, and when we have a number of running threads that is significantly greater than the number of available processors, we'll also incur in a lot of context switching cost.

This probably explains what you observed in your previous comment, because for each `future` invocation, we're creating a new thread to be executed (via ExecutorService.submit):

> I noticed some cases where parallelization does indeed add overhead (I think tests ran 5 to 10 percent slower?)

> one of the methods the most time spent in was java.lang.Thread.<init>. A lot of time was spent in java.util.concurrent.AbstractExecutorService.submit ()

One alternative we could explore instead is to create a [fixed thread pool ExecutorService](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/Executors.html#newFixedThreadPool(int)) limited to the number of [available cores](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Runtime.html#availableProcessors()).

We then can map each test to a function that runs it, and pass that to [invokeAll](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/ExecutorService.html#invokeAll(java.util.Collection)) (this works because a Clojure function implements Callable). This returns a list of Future, so we can do the `(doall (map deref futures))` on it the same way we're already doing.

We're still going to create a Callable and a Future per task, but we'll limit the number of threads created and the parallelism of execution to the available number of processors.

This should avoid the overhead of too much parallelism and hopefully result in parallel tests always running faster than serial tests.

What do you think?

Reply:
Something like
```clojure
(import '(java.util.concurrent ExecutorService Executors))

(defn task [num]
  (fn []
    (.println System/out num)
    (Thread/sleep 500)
    (* num num)))

(let [num-cpus (.availableProcessors (Runtime/getRuntime))
      executor (Executors/newFixedThreadPool num-cpus)
      data (range 100)
      tasks (map task data)
      futures (.invokeAll executor tasks)
      results (doall (map deref futures))]
  (println results))
```

Reply:
I do something similar in Splint:

```clojure

(defn pmap*
  "Efficient version of pmap which avoids the overhead of lazy-seq.

  (doall (pmap (fn [_] (Thread/sleep 100)) coll)) => 3.34 secs
  (pmap* (fn [_] (Thread/sleep 100)) coll) => 202 ms"
  [f coll]
  (let [executor (Executors/newCachedThreadPool)
        futures (mapv #(.submit executor (reify Callable (call [_] (f %)))) coll)
        ret (mapv #(.get ^Future %) futures)]
    (.shutdownNow executor)
    ret))
```

Reply:
I couldn't remember the default executor for `future`, and it appears to be `newCachedThreadPool` so it will save some on overhead for short-lived threads, but probably creates too many in other situations.

Reply:
For future reference, Clojure core's fixed thread pool uses the number of processors plus 2: https://github.com/clojure/clojure/blob/08a2d9bdd013143a87e50fa82e9740e2ab4ee2c2/src/jvm/clojure/lang/Agent.java#L49C1-L51C77